Dev/camera to camera detection transform

dmjoy opened the pull request.

Commit:
Initial (stubbed) detected object set filter for camera-to-camera transform

Commit:
Fleshed out camera-to-camera transform

Commit:
Added constructor for passing cameras directly

mleotta commented on arrows/core/transform_detected_object_set.cxx:
```diff
@@ -0,0 +1,279 @@
+// ------------------------------------------------------------------
+Eigen::Vector3d
+transform_detected_object_set::
+backproject_to_height( const kwiver::vital::camera_perspective_sptr camera,
```
This function is actually a special case of `backproject_to_plane` where plane = [0, 0, 1, 0].  So you can probably get rid of this one and just use `backproject_to_plane`, or have this function call `backproject_to_plane` to do the work.  If you are keeping this function and not making `height` a parameter I would call it `backproject_to_ground`.

mleotta commented on arrows/core/transform_detected_object_set.cxx:
```diff
@@ -0,0 +1,279 @@
+  Eigen::Vector3d img_pt_3;
+  img_pt_3 << img_pt, 1.0;
+
+  Eigen::Vector3d npt = camera->intrinsics()->as_matrix().colPivHouseholderQr().solve( img_pt_3 );
```
This will not handle radial distortion, but the `camera_intrinsics` class already has a function that does what you need.

    vector_2d npt = camera->intrinsics()->unmap(img_pt);

Then append the 1.0 to this 2D vector to make it homogeneous (3D)

mleotta commented on arrows/core/transform_detected_object_set.cxx:
```diff
@@ -0,0 +1,279 @@
+		  const Eigen::Vector4d box ) const
+{
+  // project base of the box to the ground
+  auto base_pt = Eigen::Vector2d((box(0) + box(2)) / 2, box(3));
```
FYI, `vector_2d` is a kwiver typedef for `Eigen::Vector2d`.  We tend to use that in kwiver code instead of the Eigen type directly.  Same for 3D.  It's actually `kwiver::vital::vector_2d` we often do `using namespace kwiver::vital;` in the cxx files to shorten the name.

mleotta reviewed (commented):
Here are some initial comments.  I have to leave early, but can look more later.

Commit:
Code review tweaks

- backproject_to_height renamed and calls backproject to plane the
  ground plane
- Now using the kwiver::vital namespace types

mleotta commented on arrows/core/transform_detected_object_set.cxx:
```diff
@@ -0,0 +1,264 @@
+  // project base of the box to the ground
+  auto base_pt = vector_2d((box(0) + box(2)) / 2, box(3));
+  vector_3d pc = this->backproject_to_ground( camera, base_pt );
+  vector_3d ray = pc - (-camera->rotation().matrix().transpose() * camera->translation());
```
I think you should be able to replace this line with
```c++
vector_3d ray = pc - camera->center();
```

mleotta commented on arrows/core/transform_detected_object_set.cxx:
```diff
@@ -0,0 +1,264 @@
+  Eigen::Matrix<double, 8, 2> projected_points;
+
+  // project the points
+  for (int i=0; i<8; i++) {
```
for style consistency I would put the `{` on the next line.

mleotta commented on arrows/core/transform_detected_object_set.h:
```diff
@@ -0,0 +1,99 @@
+
+  virtual vital::bounding_box<double> transform_bounding_box(vital::bounding_box<double>& bbox) const;
+
+  virtual Eigen::Vector3d backproject_to_ground(const kwiver::vital::camera_perspective_sptr camera,
```
You probably should also change the Eigen types to Kwiver typedefs here in the header to match the source file.

mleotta commented on arrows/core/transform_detected_object_set.cxx:
```diff
@@ -0,0 +1,264 @@
+
+  // loop over all detections
+  auto ie = input_set->cend();
+  for ( auto det = input_set->cbegin(); det != ie; ++det )
```
There is a simpler way in C++11.  Try:
```c++
for ( auto det : *input_set )
{
  auto out_det = det->clone();
```

or if that doesn't work

```c++
for ( auto const det : *input_set )
{
  auto out_det = det->clone();
```

mleotta reviewed (commented):
Here are a few more comments. Generally it looks pretty good.  Mostly I have feedback on style.

Also, you might want to break long lines.  I don't know that we've agreed upon a hard line length for KWIVER, but I try to keep lines less than 80 characters when possible.  There are cases where breaking lines in meaningful way is tricky, so it's okay to go over occasionally, but certainly keep the lines under 120 characters.

Commit:
Further code review tweaks

dmjoy commented:
Thanks Matt, I've made the suggested changes.  Once I verify the code works on real data, I'll create a PR for Kwiver proper.

Commit:
Updated master.txt release notes
